### ISSUE TITLE
vertical_expand in Grid layouts

### ISSUE THREAD

**Issue by ArtemSmaznov** [awesomeWM/awesome#3201]
This is the continuation of #3198 which was fixed but left an issue with vertical expand.

Really sorry about another one about the grid layout - I know it is very bad.

Output of `awesome --version`:

```
awesome v4.3-894-gcc67a5b40-dirty (Too long)
 • Compiled against Lua 5.3.6 (running with Lua 5.3)
 • API level: 4
 • D-Bus support: yes
 • xcb-errors support: no
 • execinfo support: yes
 • xcb-randr version: 1.6
 • LGI version: 0.9.2
 • Transparency enabled: yes
 • Custom search paths: no
```

**How to reproduce the issue:**

```
local wibox = require('wibox')
local gears = require('gears')
local beautiful = require('beautiful')

local icon = wibox.widget {
  {
    image = beautiful.awesome_icon,
    resize = true,
    widget = wibox.widget.imagebox
  },
  bg = beautiful.groups_bg,
  widget = wibox.container.background
}

local grid = wibox.widget {
  icon,
  icon,
  icon,
  icon,
  icon,
  forced_num_cols = 3,
  spacing       = 5,
  horizontal_expand = true,
  vertical_expand = false,
  horizontal_homogeneous = true,
  layout = wibox.layout.grid
}

grid._private.vertical_expand = false

return wibox.widget {
  grid,
  bg = beautiful.groups_bg,
  shape = function(cr, width, height)
    gears.shape.partially_rounded_rect(cr, width, height, false, false, false, false, beautiful.groups_radius)
  end,
  widget = wibox.container.background
}
```

**Actual result:**

Even though vertical_expand is set to false the icons still expand vertically up to a height of how much a single icons would take if it would fill the parent horizontally (i.e. setting `forced_num_cols` to `1` will result in the first icon being of the same height as when it was 3)

Looking at Awesome_icons:

![2020-10-20_20:24:43](https://user-images.githubusercontent.com/31800376/96634685-c3c8fc00-1312-11eb-887b-fa6db90757a7.png)

**Expected result:**

Icons should not leave empty spaces underneath and the cell should take the geometry of the child - in this case have the same width and height as the icon has a 1:1 ratio

**Comment by psychon** [comment]
Sorry for not getting back to. Also: Sorry, no spontaneous idea and not much time to look at this.
Anyone else feels like trying to understand the code for `grid:layout()`?

**Comment by actionless** [comment]
my blind guess is what instead of 

```lua
        -- Recalculate the height so the last widget fits
        if (fill_space or self._private.vertical_expand) and y + h > height then
            h = math.floor(math.max(self._private.min_rows_size, height - y))
        end
```

it should be 

```lua
        if (fill_space or self._private.vertical_expand) and y + h > height then
            h = math.floor(math.min(self._private.min_rows_size, height - y))
        end
```

but i'm lazy to adapt reporter's example for running with default rc.lua

**Comment by actionless** [comment]
@ArtemSmaznov have you tried setting `vertical_homogeneous` property to `false`?

could you also adapt your example as a patch to default rc.lua so i could run it easily?

**Comment by ArtemSmaznov** [comment]
Quite sure I tried setting homogeneous as you suggested, and yeah I will adapt the example to the default rc no probs

**Comment by ArtemSmaznov** [comment]
@actionless yes - definitely tried `vertical_homogeneous=false`

here is the example adapted for the default rc.lua

```
local icon = wibox.widget {
  {
    image = beautiful.awesome_icon,
    resize = true,
    widget = wibox.widget.imagebox
  },
  bg = '#ccc',
  widget = wibox.container.background
}

local grid = wibox.widget {
  icon,
  icon,
  icon,
  icon,
  icon,
  forced_num_cols = 3,
  spacing       = 5,
  horizontal_expand = true,
  vertical_expand = false,
  horizontal_homogeneous = true,
  layout = wibox.layout.grid
}

grid._private.vertical_expand = false

local wi = wibox.widget {
  grid,
  bg = beautiful.groups_bg,
  shape = function(cr, width, height)
    gears.shape.partially_rounded_rect(cr, width, height, false, false, false, false, beautiful.groups_radius)
  end,
  widget = wibox.container.background
}

local w = wibox{ x = 100, y = 50, width = 300, height = 600, visible = true, widget = wi }
```

**Comment by actionless** [comment]
i revisited what happens in grid's layout and all seems legit:

1)your `icon` widget doesn't have any size (it's resizeable, right?)

2) and there's no option in grid like "follow aspect ratio"

so those two things lead to result what it tried to build an optimal grid

you can solve that by defining dimensions of your "sub-widget", for example:

```
local icon = wibox.widget {
  {
    {
      image = beautiful.awesome_icon,
      resize = true,
      widget = wibox.widget.imagebox
    },
    bg = '#ccc',
    widget = wibox.container.background
  },
  height   = 128,
  layout = wibox.container.constraint,
}
```

**Comment by actionless** [comment]
or, to rephrase my previous message, when sub-widget doesn't have any dimensions, layout logic could handle it only as it's always set to *_expand = true, since there is no other source of knowledge for it's size

**Comment by ArtemSmaznov** [comment]
My goal is to make a grid of apps that is independent of the size of the parent widget (in my case a notification tray), which would be of different sizes on different monitors. As such I do not know what would be actual dimensions of widgets should be.

I want widget to display a set number of sub-widgets of equal sizes in a row and when there are more widgets available move to the next row. Grid layout sounded like it was exactly what I needed but those gaps just don't make much sense to me and I cannot figure out a way to achieve this without hardcoding specific geometry for the sub-widgets. As the sub-widget gets constrained by it's width by the grid layout I expected each row of the grid to be as high as the widget gets. In case of square icons for the cells to be squares as well.

If there is another approach I can take to achieve this I would be glad to hear it.

Just to give one of the examples I am working on. I have a side panel that would be of different sizes depending on the display it is on. That panel has a list of frequent apps that is using a grid layout and sub-widgets (icons) with set max_height and max_width that look fine for my main monitor. But on another screen or another computer or even if I decide to change the size of the panel all the icons start to look bad as there is less/more space to house items of a fixed size correctly.

![2020-10-21_16:41:02](https://user-images.githubusercontent.com/31800376/96744165-a9e3f380-13bc-11eb-93d4-46025e4cedb4.png)

**Comment by actionless** [comment]
> with set max_height and max_width

that part is missing in your minimal example

**Comment by ArtemSmaznov** [comment]
max_height and max_width is a workaround I use right now to achieve the desired result and is what I am trying to get away from.
My expectation is that the grid_layout should produce the same result without them. At least that is what I am trying to acomplish

**Comment by actionless** [comment]
i'm not sure on what this expectation could be based, taking into consideration the details i've mentioned above

**Comment by actionless** [comment]
i could imagine the result you want to achieve be possible without explicitly setting height only if in grid layout would be implemented some new property like "honor aspect ratio"

**Comment by ArtemSmaznov** [comment]
I see. Guess the use case for grid layout is something else than what I imagined then.

Thanks for the time. Think I need to write a custom function that would build the grid and pass on the parent widget'd dimensions into it so I can use those to calculate the dimensions for the sub-widgets.

Just wanted to make sure that vertical expand is not a bug of the grid layout. After web design it feels weirdly complicated working with how awesome widgets are later out and I most probably just don't get the purpose of the grid layout.

We can close this I guess

**Comment by actionless** [comment]
i'll let it open for either @psychon or @Elv13 to have a second eye on it

**Comment by psychon** [comment]
> i'll let it open for either @psychon or @Elv13 to have a second pair of eyes on it

I would really like to avoid having to understand `grid.lua`... Instead:

> Think I need to write a custom function

I'm not quite sure what you have in mind here, but I gave it a try. The following behaves quite badly when the height is too small for everything to fit (placing widgets outside of the bound of the grid layout) and also I am not quite sure what to do in `:fit()`, but besides that it seems to work for me:
![foo](https://user-images.githubusercontent.com/89482/96888711-327a9680-1486-11eb-8f38-cea95a0990d7.png)
```lua
local icon = wibox.widget {
  {
    image = beautiful.awesome_icon,
    resize = true,
    widget = wibox.widget.imagebox
  },
  bg = '#ccc',
  widget = wibox.container.background
}

local function my_grid(num_cols, spacing)
    local children = {}
    local self = wibox.widget.base.make_widget()
    function self:set_children(new_childs)
        children = new_childs
        self:emit_signal("widget::layout_changed")
    end
    function self:fit(context, width, height)
        -- We can compute the number of rows that we need:
        local num_rows = 1 + math.ceil((#children - 1) / num_cols)
        if num_rows == 0 then
            return 0, 0
        end
        -- Let's assume each child gets a quadratic area. Then we can compute a
        -- size for "things" based on the available width. This goes seriously
        -- wrong when the height is too small, since we are ignoring that here.
        local size = (width - spacing * (num_cols - 1)) / num_cols
        return width, size * num_rows + spacing * (num_rows - 1)
    end
    function self:layout(context, width, height)
        local num_rows = 1 + math.ceil((#children - 1) / num_cols)
        if num_rows == 0 then
            return {}
        end
        local size = (width - spacing * (num_cols - 1)) / num_cols
        local result = {}
        for index, child in ipairs(children) do
            local x = (index - 1) % num_rows
            local y = math.floor((index - 1) / num_rows)
            print(string.format("%d at (%d, %d) for %d rows", index, x, y, num_rows))
            local x, y = x * (size + spacing), y * (size + spacing)
            table.insert(result, wibox.widget.base.place_widget_at(child, x, y, size, size))
        end
        return result
    end
    return self
end

local grid = wibox.widget {
  icon,
  icon,
  icon,
  icon,
  icon,
  forced_num_cols = 3,
  spacing       = 5,
  horizontal_expand = true,
  vertical_expand = false,
  horizontal_homogeneous = true,
  layout = wibox.layout.grid
}

grid._private.vertical_expand = false

local wi = wibox.widget {
  grid,
  bg = beautiful.groups_bg,
  shape = function(cr, width, height)
    gears.shape.partially_rounded_rect(cr, width, height, false, false, false, false, beautiful.groups_radius)
  end,
  widget = wibox.container.background
}

local w = wibox{ x = 100, y = 50, width = 300, height = 600, visible = true, widget = wi }

local grid = my_grid(3, 5)
grid:set_children{ icon, icon, icon, icon, icon }
local wi = wibox.widget {
  grid,
  bg = beautiful.groups_bg,
  shape = function(cr, width, height)
    gears.shape.partially_rounded_rect(cr, width, height, false, false, false, false, beautiful.groups_radius)
  end,
  widget = wibox.container.background
}

local w = wibox{ x = 450, y = 50, width = 300, height = 600, visible = true, widget = wi }
```

**Comment by ArtemSmaznov** [comment]
@psychon 
I was thinking of trying to set all the sizes globally in a separate file and try to pass the size of the parent in the grid that way but will surely give your function a try.

Would like your opinion thought on if you think the sub-widget expanding vertically even if it is restricted horizontally is something that should be happening. I get what @actionless is saying about the logic behind it but I still feel like something like a simple grid should be achieved easier without the need for such complex workarounds.

**Comment by actionless** [comment]
> the sub-widget expanding vertically

it's not "expanding" - it's just **don't have size**

**Comment by ArtemSmaznov** [comment]
But they only go as high as a single icon would go. They don't stretch to the whole screen vertically. I just don't why as the icons height is not so large anymore.

**Comment by psychon** [comment]
Sigh, okay, so I'll take another look at the grid layout...

The layout asks its sub-widgets for their size. Imageboxes say "give me everything you got" (because images are scalable). In default mode every cell gets the maximum of the size that any child asked for (which is why you see only one imagebox - they asked for all the available space). With horizontal/vertical expand, it instead calculates the size for each widget based on the configured number of colums/rows.

There is also a homogeneous setting... uhm... that just scales things down? It looks at how much space all the widgets together asked for and then scales things down to keep the ratio between widgets until they fit into the available space.

Anyway, I guess the problem comes down to "images do not have a natural size (unlike e.g. a textbox) so they just ask for all the available space". Thanks to that, the grid layout does not really have much to work with. The best idea that comes to mind right now would be to preserve aspect ratios when scaling widgets down, but that only works for imageboxes. Textboxes end up being weird if you do that.

> But they only go as high as a single icon would go. They don't stretch to the whole screen vertically. I just don't why, as the icon's height is not so large anymore.

Let's walk through an example.

The example will be a grid layout with five children, each of which is a 10x10 image. The grid layout is configured to have three columns.
Let's assume that the parent layout assigned the grid layout an area of 300x300 pixels and it has to do something with this. The grid layout now asks each child "I got 300x300 pixels, how much of that do you want?" and each imagebox says "everything!". So, this makes a total width of 900 pixels (because we have three columns) and a total height of 600 pixels (because we need two rows to make the five children fit).

Okay, now without `expand=false`, the first widget gets the 300x300 pixels and the other widgets get the remaining 0x0 pixels.

With `expand=true`... well, we have a total "request" for 900x600 pixels. To make that fit into 300x300 pixels, we have to divide the widths by 3 and the heights by 2. So, each widget is assigned an area of 100x150 pixels and has to make the best use of this.
I guess the name "expand" makes more sense when the widgets asked for less space than is available and they are scaled up...

`homogeneous=true` does not make a difference here. It comes into play when the children ask for different amount of space. With `homogeneous=true`, all cells must have the same size while without it, it is possible to have different widths for the different rows/heights. Not relevant here.

I hope this explanation helps a bit.

**Comment by ArtemSmaznov** [comment]
@psychon yeah that all makes sense but how the following will fit into this.
1. Let say we have five child images that are set to expand (take all available space)
1. And we have a grid with an area of `300x1000` with `forced_cols=3`
1. When we set `horizontal_homogeneous=true` the grid limits the width of each image to `300/3=100` and shows 3 images of equal sizes in a row (this is what is actually happening after your last fix)
1. now the images are scaled down from max size to be `100x100` (even though it is not hardcoded)
1. when the grid calculates the height of cells it doesn't pick up that the image is now forced into `100x100` size and still picks up that the image would have been `300x300` as if `horizontal_homogeneous` was set to `false`
1. this makes the cell take the width of `300/3=100` and height of `300`.
1. with 5 widgets it will result in the grid layout taking horizontal space correctly (or shall I say intuitively) but the vertical space is divided into 2 rows with the `height` of `300` each and 400 remaining empty space. (empty remaining space seems to be a behavior I would expect)

What I feel should work differently is point 5 as in the image is forced into a particular size and the vertical cell size should account for that. Because if the image was requesting all available space vertically then the grid would not limit cell's height to `300` nor leave empty space in the end.

I understand that this might not be possible due to the way the layout is written or how sizes work. I simply want to point out what seems like a buggy behavior.

**Comment by actionless** [comment]
> the image is forced into a particular size and the vertical cell size should account for that.

> i could imagine the result you want to achieve be possible without explicitly setting height only if in grid layout would be implemented some new property like "honor aspect ratio"

**Comment by actionless** [comment]
alternatively, it could be some new property like "2-dimensional homogenous" (:-) or some nicer name)

but my point what the current implementation/documentation doesn't give any clues to support your current assumption

**Comment by ArtemSmaznov** [comment]
I didn't get that at the time you first mentioned it but now I think yeah. That might be what I'm after

**Comment by ArtemSmaznov** [comment]
> but my point what the current implementation/documentation doesn't give any clues to support your current assumption
I'm not arguing that it does. Just trying to get the grid layout to be more versatile I guess

**Comment by actionless** [comment]
i'm really glad it helped :-)

in that case feel free to submit PR for implementing one of those properties

**Comment by psychon** [comment]
@ArtemSmaznov well... let's imagine a weird "non-image-like" widget: A textbox that shows text in multiple lines and only ellipsizes whole words.

When given enough space, it would show "The quick brown fox jumps over something". When the width gets smaller, the height of this widget increases because the text gets "squished" into two lines. In this situation, it can happen that you offer the widget 300 pixels of width, but it only wants 250 (because that's the end of the last word that fits into the line).

This widget has "weird" size behaviour. And the only API that we have for working with this is "if I offer you this much space, how much of it do you want to use?".

> What I feel should work differently is point 5 as in the image is forced into a particular size and the vertical cell size should account for that.

Okay, let's say the current grid layout is modified. It does what it does now. When it is done, it re-queries the widgets for their sizes instead of directly assigning them the computed sizes. What does it do then? Repeat everything, then query again with the newly computed sizes and repeat this until nothing changes any more? Nothing guarantees that this does not result in an endless loop. Also, I bet this can result in non-intuitive results.

Basically, what I am trying to say is similar to what you said before:

> I understand that this might not be possible due to the way the layout is written or how sizes work.

I'm not really saying this this is impossible, but at least I do not really see how this is supposed to work...

Random idea: We could make a vertical/horizontal grid layout. It is assigned a number of columns/rows, uses that to compute the space for each widget (based on the available space) and then asks the widgets for their size in "the other direction" (for a horizontal grid, this would then compute the height of the widget). I think this would be similar to what you are asking for, but it can only do what the current grid layout calls "homogeneous" (I am sure I misspelt that...).

So... dunno. Grids are complicated.

Thanks for staying with us even though I have the impression we only ever say "no, that's impossible". Let's hope this ends up in some improvements to AwesomeWM. :-)

**Comment by ArtemSmaznov** [comment]
@psychon I see where my logic was flawed. I assumed the height of cells was calculated after the width but I see now that those 2 parameters get defined at the same time (sorry if I am putting it in the wrong way but I thing you get what I am saying).

If the `height` was calculated (in case of a vertical grid) after the `width` then the text widget example you provided would work fine as well. I imagine that would entail rewriting pretty much the whole logic for the grid layout though and is not something I could reasonably expect to be done.

What do you think of the optional `respect aspect ratio` option? Is that something that can be done relatively easy? I have absolutely no idea how one might go about implementing that. You obviously have a much better understand of it as I have only dealt with the custom rc.lua so far. Maybe one day I will get into contributing to the actual source code.

And no worries. I did not really expect to get much of a change with this. Just wanted to point out what I found weird and see where it takes us. I did enjoy this conversation and only hope I didn't take too much of both your time. I do enjoy AwesomeWM and only wanna see it get better. I do find it quite hard to configure and understand sometimes but think I am getting better.

**Comment by psychon** [comment]
> If the height was calculated (in case of a vertical grid) after the width then the text widget example you provided would work fine as well.

Yup. So I guess, the best way to fix this issue is to write another grid layout that works differently than the existing one, for example in the way you are suggesting here. However, no idea how to document this so that things make sense (or even how to call it). I would guess that such a new mode of operation does not fit into the existing grid layout (which was written 3.5 years ago and did not see much modifications since then).
